qbft/spectest/tests/controller/processmsg/out_of_order_flow.go

```diff
@@ -0,0 +1,75 @@
+package processmsg
+
+import (
+	"github.com/bloxapp/ssv-spec/qbft"
+	"github.com/bloxapp/ssv-spec/qbft/spectest/tests"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/bloxapp/ssv-spec/types/testingutils"
+	qbftcomparable "github.com/bloxapp/ssv-spec/types/testingutils/comparable"
+)
+
+// OutOfOrderFlow tests a QBFT execution with messages out of order
+// One prepare and one commit are received before the proposal causing the instance not to decide
+func OutOfOrderFlow() tests.SpecTest {
+	ks := testingutils.Testing4SharesSet()
+
+	correctFlowMessgaes := testingutils.DecidingMsgsForHeightWithRoot(testingutils.TestingQBFTRootData,
+		testingutils.TestingQBFTFullData, testingutils.TestingIdentifier, qbft.FirstHeight, ks)
+
+	// shuffles messages
+	outOfOrderMessages := []*qbft.SignedMessage{
+		correctFlowMessgaes[3], // prepare 3
+		correctFlowMessgaes[6], // commit 3
+		correctFlowMessgaes[0], // proposal 1
+		correctFlowMessgaes[1], // prepare 1
+		correctFlowMessgaes[2], // prepare 2
+		correctFlowMessgaes[4], // commit 1
+		correctFlowMessgaes[5], // commit 2
+	}
+
+	sc := OutOfOrderFlowStateComparison(outOfOrderMessages)
+
+	return &tests.ControllerSpecTest{
+		Name: "out or order flow",
+		RunInstanceData: []*tests.RunInstanceData{
+			{
+				InputValue:    testingutils.TestingQBFTFullData,
+				InputMessages: outOfOrderMessages,
+				ExpectedDecidedState: tests.DecidedState{
+					DecidedVal: nil,
+				},
+				ControllerPostRoot:  sc.Root(),
+				ControllerPostState: sc.ExpectedState,
+			},
+		},
+		ExpectedError: "could not process msg: invalid signed message: did not receive proposal for this round",
```

Comment:
You know, until this day this behavior in spec confuses me...
Technically, messages on a real network may get delayed and come out of order.

Also looking again at the IBFT paper it seems that out of order messages are allowed to my understanding...

To my understanding the implementation uses the assumption that messages come in order to increase performance. This may be fine and maybe we adjusted the spec to work this way.

But it is kind of weird that we restrict out of order messages... maybe it is best to not have a test for it so that implementation can choose their behavior?

@moshe-blox @MatheusFranco99

Reply:
I agree that it's a confusing behaviour.
IBFT totally allows it while QBFT implementation doesn't, as it seems to me.
In my opinion this could change. We could store messages that seems valid and only process the IBFT rule when something triggers it (e.g. storing all prepares and, once a proposal is received, we check the IBFT rule for that proposal).

Reply:
@moshe-blox or @y0sher 

just curious if you ever saw duties fail/delay due to out of order messages

Reply:
currently I don't think we should merge this because why enforce this?

Reply:
I agree. I don't really see the point of this test.

Reply:
Basically, we are talking about processing messages the way you said, we did see sometimes where we get a prepare before propose and then fail to process the propose. we want to create a structure like you mentioned where we store valid looking messages that arrive and process them in the appropriate time instead of just crashing because of it.

Reply:
I'm not reviewing this (although it looks fine) for now.

Reply:
Even if eventually you will do this @y0sher 
not sure this test should be merged because I am unsure if we should force implementation to accept out of order messages (even though theoretically we should)

Reply:
Agree